### PR TITLE
Fix backwards `cmpxchg_weak_failure_rate` check

### PR DESCRIPTION
**Pull request by thomcc**
Uh, I guess I got this backwards a long time ago 😓.

### CODE REVIEW

File: src/data_race.rs [hunk 1]
```diff
@@ -597,7 +597,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         // of the time, based on `rate`.
         let rate = this.machine.cmpxchg_weak_failure_rate;
         let cmpxchg_success = eq.to_scalar()?.to_bool()?
-            && (!can_fail_spuriously || this.machine.rng.get_mut().gen::<f64>() < rate);
+            && (!can_fail_spuriously || this.machine.rng.get_mut().gen::<f64>() >= rate);
```
Comment (RalfJung) [inline src/data_race.rs:600]: Please add a comment explaining `>=` vs `>`: `gen` returns a value from the half-open range `[0, 1)`, so `rate == 1` is guaranteed to lead to a `false` here.

**Comment by cbeuw** [comment]
Ah this is why crossbeam had to set `-Zmiri-compare-exchange-weak-failure-rate=1.0` to fix Miri hanging on tests https://github.com/crossbeam-rs/crossbeam/pull/578#issuecomment-782555221, which made no sense. @taiki-e would need to change that back to `-Zmiri-compare-exchange-weak-failure-rate=0.0` (I pinned down why this was needed in the first place, will post it in a separate crossbeam issue)

File: src/data_race.rs [hunk 2]
```diff
@@ -597,7 +597,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         // of the time, based on `rate`.
         let rate = this.machine.cmpxchg_weak_failure_rate;
         let cmpxchg_success = eq.to_scalar()?.to_bool()?
```
Comment (cbeuw) [inline src/data_race.rs:599]:
```suggestion
        let success_rate = 1.0 - this.machine.cmpxchg_weak_failure_rate;
        let cmpxchg_success = eq.to_scalar()?.to_bool()?
            && if can_fail_spuriously {
                this.machine.rng.get_mut().gen::<f64>() < success_rate
            } else {
                true
            };
```
I think this is clearer? We were assigning to a success bool using a failure rate so we should invert the rate first, also implications should be written out as `if` expressions

Comment (RalfJung) [reply src/data_race.rs:599]: I agree about the implication; no strong opinion about using `>= fail_rate` or `< 1-fail_rate`.

Comment (cbeuw) [reply src/data_race.rs:599]: Actually we should just use [`gen_bool()`](https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen_bool)

Comment (thomcc) [reply src/data_race.rs]: Sorry for the delay. I've switched to use gen_bool, and used the suggested change, but can't run tests/checks locally at the moment, so maybe there's some issue.

File: src/data_race.rs (hunk 1 above)
Comment (thomcc) [reply src/data_race.rs]: I think use of gen_bool makes it clear enough? Maybe?

**Comment by RalfJung** [comment]
This reminds me, is there some test we could add for this?

**Pull request merged**